### DISCUSSION TITLE
PorousFlow Unexpected Mass Balance Results

### DISCUSSION THREAD

**Discussion by jrpatterson79** [idaholab/moose#26059] [Q&A Modules: Porous Flow]
Hello MOOSE Community,

I'm de-bugging a model that simulates fluid injection through a 2D fracture embedded in a 3D volume, using a conforming mesh built in GMSH and converted to EXODUS format using MOOSE. 

I'm using [PorousFlowSquarePulsePointSource](https://mooseframework.inl.gov/source/dirackernels/PorousFlowSquarePulsePointSource.html) block restricted to the fracture subdomain to simulate constant fluid injection at the center of the fracture. I've prescribed a large permeability contrast between the fracture and rock subdomains such that fluids should move along the fracture, with minimal to no fluid exchange into the surrounding rock domain.

Given this, I would expect the mass balance block restricted to the fracture subdomain should be abs(fluid_in - fluid_out) = inj_rate * dt
However, the fracture mass balance is a little more than an order-of-magnitude less than injected mass per time step. In contrast, the block restricted mass balance in the rock subdomain is equivalent to the injected mass per time step. This makes me concerned that fluid is not being injected into the fracture subdomain, but instead into the rock.

Is this expected behavior / something that I should not be concerned with? Is this the result of an error in meshing / input file on my part?

I have a MWE at https://github.com/jrpatterson79/testing.git

Thanks in advance for any insights y'all might have
JPatterson

**Comment by lynnmunday** [comment]
@1runer

**Reply by 1runer** [reply]
I hope @cpgr and @WilkAndy have a better overview about this.

**Reply by GiudGiud** [reply]
For rock_mass_blnc you could use a ParsedPostprocessor. Using a function is a little hacky as the dependency resolution wont be done properly. 

Does this imbalance remain with finer discretizations?

**Reply by cpgr** [reply]
I suspect that the issue is discretisation and the dirac point at a node shared by the fracture and rock matrix means that the injected fluid is already in the matrix. If you make the rock porosity vanishingly small then it probably reduces this affect.

**Reply by jrpatterson79** [reply]
@GiudGiud thank you for pointing out the ParsedPostprocessor, I was not aware of that. I've made that switch in my input file. To answer your question, this behavior has been consistent moving from a coarse mesh to the currently more refined mesh. Refining the mesh even produces a minimal effect. 

@cpgr Reducing the porosity does diminish the effect, which makes sense. I've spent sometime re-reading the [flow through fractured media page](https://mooseframework.inl.gov/modules/porous_flow/nomultiapp_flow_through_fractured_media.html), and I'm wondering if there is an effect of the element shape / type. GMSH produces TRI3 elements for the 2D fracture surface. In contrast, the 3D example (coarse_3D.i) shows that the 2D fracture surfaces are TRISHELL3 elements, which appears to be an element type generated by CUBIT. Could this be contributing to the results I'm seeing? Is GMSH capable of generating this type of element? I have not found any helpful answers through internet searches or reading through the GMSH user manual

**Reply by lynnmunday** [reply]
I use TRI3 for 2D fracture elements, that is not problem.

**Comment by lynnmunday** [comment]
Does your dirac kernel produce an error that the point is not found?  Also, you are restricting the dirac kernel to the fracture block so the input should limited to the 2D fracture element and should not be smeared out over the 3D to the matrix element.  Can you check that?  You can check with vector tags to see if the dirac kernel is injecting in the matrix, use something like this:

``` 
[Problem]
  extra_tag_vectors = 'flux_tag'
[]

[AuxVariables]
  [flux_tag]
    order = FIRST
    family = LAGRANGE
  []
[AuxKernels]
  [flux]
    type = TagVectorAux
    variable = flux_tag
    v = pp
    vector_tag = flux_tag
    execute_on = timestep_end
  []
[]

# Hydraulic injection kernel
[DiracKernels]
  [mass_flux_source]
    type = PorousFlowSquarePulsePointSource
    mass_flux = ${injection_rate}
    point = '0 0 175'
    variable = pp 
    block = 'fracture'
    point_not_found_behavior = WARNING
    extra_vector_tags = 'flux_tag'
  []
[]

**Reply by jrpatterson79** [reply]
@lynnmunday 

The Dirac kernel gives a point not found warning regardless of which block I assign it to or leave it unrestricted. Looking at nodelist 1 in the exodus mesh file, the point '0, 0, 175' exists, so it's not clear to me why that error is produced.

Adding the TagVectorAux and looking at the exodus results file, the flux_tag = -inj_rate in the fracture subdomain at the point '0, 0, 175'

**Comment by cpgr** [comment]
One thing that struck me is that the mass of water in the fracture is reported as ~ 3,600 kg, but the fracture looks like a circular disk with a radius of 45 m and aperture of 3e-4 m, so volume ~ 2 m3, which should only contain ~2,000 kg of water... unless I'm missing something?

**Reply by jrpatterson79** [reply]
@cpgr - Yes, that is all correct. One thing I will double-check tomorrow is that there is not an additional side set with the name ‘fracture’ that is being added into the porousflowmass postprocessor. 

Any other ideas of things I should double check?
